exporters/prometheus/src/main/java/io/opentelemetry/exporters/prometheus/MetricAdapter.java

```diff
@@ -0,0 +1,192 @@
+import io.opentelemetry.sdk.metrics.data.MetricData.ValueAtPercentile;
+import io.prometheus.client.Collector;
+import io.prometheus.client.Collector.MetricFamilySamples;
+import io.prometheus.client.Collector.MetricFamilySamples.Sample;
+import io.prometheus.client.Collector.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Util methods to convert OpenTelemetry Metrics data models to Prometheus data models.
+ *
+ * <p>Each OpenTelemetry {@link MetricData} will be converted to a Prometheus {@link
+ * MetricFamilySamples}, and each {@code Point} of the {@link MetricData} will be converted to
+ * Prometheus {@link Sample}s.
+ *
+ * <p>{@code DoublePoint}, {@code LongPoint} will be converted to a single {@link Sample}. {@code
+ * Summary} will be converted to two {@link Sample}s (sum and count) plus the number of Percentile
+ * values {@code Sample}s
+ *
+ * <p>Please note that Prometheus Metric and Label name can only have alphanumeric characters and
+ * underscore. All other characters will be sanitized by underscores.
+ */
+final class MetricAdapter {
+
+  static final String SAMPLE_SUFFIX_COUNT = "_count";
+  static final String SAMPLE_SUFFIX_SUM = "_sum";
+  static final String LABEL_NAME_QUANTILE = "quantile";
+
+  // Converts a MetricData to a Prometheus MetricFamilySamples.
+  static MetricFamilySamples toMetricFamilySamples(MetricData metricData) {
+    Descriptor descriptor = metricData.getDescriptor();
+    String fullName =
+        toMetricFullName(
+            descriptor.getName(), metricData.getInstrumentationLibraryInfo().getName());
+    Type type = toMetricFamilyType(descriptor.getType());
+
+    return new MetricFamilySamples(
+        fullName,
+        type,
+        descriptor.getDescription(),
+        toSamples(fullName, descriptor, metricData.getPoints()));
+  }
+
+  private static String toMetricFullName(
+      String descriptorMetricName, String instrumentationLibraryName) {
+    if (instrumentationLibraryName.isEmpty()) {
+      return Collector.sanitizeMetricName(descriptorMetricName);
+    }
+
+    // Use "_" here even though the right way would be to use "." in general, but "." will be
+    // replaced with "_" anyway so one less replace call.
+    return Collector.sanitizeMetricName(instrumentationLibraryName + "_" + descriptorMetricName);
+  }
+
+  static Type toMetricFamilyType(MetricData.Descriptor.Type type) {
+    switch (type) {
+      case NON_MONOTONIC_LONG:
+      case NON_MONOTONIC_DOUBLE:
+        return Type.GAUGE;
+      case MONOTONIC_LONG:
+      case MONOTONIC_DOUBLE:
+        return Type.COUNTER;
+      case SUMMARY:
+        return Type.SUMMARY;
+    }
+    return Type.UNTYPED;
+  }
+
+  // Converts a list of points from MetricData to a list of Prometheus Samples.
+  static List<Sample> toSamples(String name, Descriptor descriptor, Collection<Point> points) {
```

Comment:
this method is gigantic. Would be good to refactor it into smaller chunks.

Reply:
A bit of splitting :)